database/mongodb/mongodb.go

```diff
@@ -184,10 +207,70 @@ func (m *Mongo) Drop() error {
 	return m.db.Drop(context.TODO())
 }
 
+func (m *Mongo) ensureLockTable() error {
+	indexes := m.db.Collection(m.config.LockingCollection).Indexes()
+	indexOptions := options.Index().SetUnique(true).SetName("lock_unique_key")
+	indexKeys := bsonx.MDoc{
+		DefaultLockingKey: bsonx.Int32(-1),
+	}
+	_, err := indexes.CreateOne(context.TODO(), mongo.IndexModel{
+		Options: indexOptions,
+		Keys:    indexKeys,
+	})
+	if err != nil {
+		fmt.Println(err)
+		return err
+	}
+	return nil
+}
+
+// ensureVersionTable checks if versions table exists and, if not, creates it.
+// Note that this function locks the database, which deviates from the usual
+// convention of "caller locks" in the MongoDb type.
+func (m *Mongo) ensureVersionTable() (err error) {
+	if err = m.Lock(); err != nil {
+		return err
+	}
+
+	defer func() {
+		if e := m.Unlock(); e != nil {
+			if err == nil {
+				err = e
+			} else {
+				err = multierror.Append(err, e)
+			}
+		}
+	}()
+
+	if err != nil {
+		return err
+	}
+	if _, _, err = m.Version(); err != nil {
+		return err
+	}
+	return nil
+}
+
 func (m *Mongo) Lock() error {
+	newLockObj := lockObj{
+		Key: lockKey,
+	}
+	_, err := m.db.Collection(m.config.LockingCollection).InsertOne(context.TODO(), newLockObj)
+	if err != nil {
+		return database.ErrLocked
+	}
```

Comment:
So one thing to note, is that the first thread that comes up, can get the lock because of the ensureVersionTable(), then another thread that comes can get the lock because of the ensureVersionTable(), and then the first thread can _fail_ to get the lock and return an error (I think that's what happens now), because of the second thread.